src/lsst/rsp/catalog.py

```diff
@@ -70,20 +87,82 @@ def retrieve_query(query_url: str) -> pyvo.dal.AsyncTAPJob:
+        discovery information within a Nublado notebook.
+
+    Returns
+    -------
+    list[str]
+        A list of strings in the format dataset:query_id.
+
+    Notes
+    -----
+    This formerly assumed "/api/tap"; it no longer does.  The way we're
+    handling this, and the transition between the two, is that we will do the
+    last n jobref ids across all datasets found in the discovery document.
+
+    What we will return is a list of strings (still) where the string returned
+    is dataset:query_id.  Then the caller must check whether there is a
+    colon in the returned value, and use that to resolve the correct endpoint
+    to send the query to.  If no colon exists, it's at "/api/tap".
     """
-    client = RSPClient("/api/tap")
-    params = {}
-    if n and n > 0:
-        params = {"last": f"{n}"}
-    full_history_xml = await client.get("async", params=params)
-    history_dict = xmltodict.parse(
-        full_history_xml.text, force_list=("uws:jobref",)
-    )
-    try:
-        joblist = history_dict["uws:jobs"]["uws:jobref"]
-    except KeyError:
-        return []
-    return [job["@id"] for job in joblist if "@id" in job]
+    datasets = list_datasets(discovery_v1_path=discovery_v1_path)
+    # Some datasets share an endpoint.  In that case, we only want to get
+    # the n most recent for a given endpoint, so we do some deduplication
+    # here.  (So, for instance, if dp1 and dp02 share "/api/tap", what we
+    # want is the n most recent from both of those put together, not the n
+    # most recent from each.)
+    #
+    # It won't matter in that case whether we assign the query as having
+    # belonged to dp1 or dp02, because it's the endpoint:query_id tuple
+    # that is unique.
+    client: dict[str, RSPClient] = {}
+    seen_endpoints: set[str] = set()
+    for ds in datasets:
+        url = get_service_url("tap", ds)
+        if url:
+            if url not in seen_endpoints:
+                seen_endpoints.add(url)
+                client[ds] = RSPClient(url)
+
+    params: dict[str, str] = {}
+    params = {"last": f"{n}"} if n and n > 0 else {}
+    resp: dict[str, httpx.Response] = {}
+    history: dict[str, Any] = {}
+    qlist: list[dict[str, Any]] = []
+    for ds, ds_client in client.items():
+        resp[ds] = await ds_client.get("async", params=params)
+        history[ds] = xmltodict.parse(
+            resp[ds].text, force_list=("uws:jobref",)
+        )
+
+    for ds, history_ds in history.items():
+        if "uws:jobs" in history_ds:
+            if "uws:jobref" in history_ds["uws:jobs"]:
+                for entry in history_ds["uws:jobs"]["uws:jobref"]:
+                    # Annotate with dataset name
+                    entry["__uws:dataset"] = ds
+                    qlist.append(entry)
+
+    # It appears that uws:creationTime can be lexically sorted and will do
+    # the right thing.
+    qlist.sort(key=lambda x: x["uws:creationTime"], reverse=True)
```

Comment:
Jobs should indeed always have a `creationTime`, but if theoretically one didn't this would break.
Since it should always be there it is fine to leave it as is, but if you want to be super defensive you could also add a check here.